rmqproducer/rabbitmq_producer.py

```diff
@@ -0,0 +1,390 @@
+        on_delivery_confirmation method will be invoked passing in a Basic.Ack
+        or Basic.Nack method from RabbitMQ that will indicate which messages it
+        is confirming or rejecting.
+
+        """
+        self._LOGGER.info('Issuing Confirm.Select RPC command')
+        self._channel.confirm_delivery(self.on_delivery_confirmation)
+        self._connection.ioloop.stop()
+
+    def on_delivery_confirmation(self, method_frame):
+        """Invoked by pika when RabbitMQ responds to a Basic.Publish RPC
+        command, passing in either a Basic.Ack or Basic.Nack frame with
+        the delivery tag of the message that was published.
+
+        The delivery tag is an integer counter indicating the message number
+        that was sent on the channel via Basic.Publish. Here we're just doing
+        house keeping to keep track of stats and remove message numbers that
+        we expect a delivery confirmation of from the list used to keep track
+        of messages that are pending confirmation.
+
+        :param pika.frame.Method method_frame: Basic.Ack or Basic.Nack frame
+
+        """
+        confirmation_type = method_frame.method.NAME.split('.')[1].lower()
+        message_num = method_frame.method.delivery_tag
+        if confirmation_type == 'ack':
+            self._LOGGER.info('Message %i published successfully',
+                              message_num)
+            self._LOGGER.debug('The message published: %s',
+                               self._messages[message_num]['message'])
+        if confirmation_type == 'nack':
+            self._LOGGER.error('The message %i failed to publish: %s',
+                               message_num, self._messages[message_num]['message'])
+            if self.nack_callback:
+                self.nack_callback(self._messages[message_num]['message'])
+        del self._messages[message_num]
+        self._connection.ioloop.stop()
+
+    def publish_message(self, message, routing_key):
+        """If the class is not stopping, publish a message to RabbitMQ,
+        appending a list of deliveries with the message number that was sent.
+        This list will be used to check for delivery confirmations in the
+        on_delivery_confirmations method.
+
+        :param str message: The message to be published
+        :param str routing_key: The routing key for the message to be published
+
+        """
+        if self._channel.is_open:
+            self._channel.basic_publish(self.exchange, routing_key, message)
+        else:
+            self._LOGGER.error("Channel not open. Message %s couldn't be published. "
+                               "Will try to publish message again if channel reopens", message)
+        self._message_number += 1
+        self._messages[self._message_number] = {
+            'message': message, 'routing_key': routing_key}
+        self._LOGGER.debug('Publishing message # %i', self._message_number)
+        self._connection.ioloop.start()
```

Comment:
ioloop can't be discarded. After the line 'self._connection.ioloop.start()' code will indefinitely fall into ioloop untill the ioloop is stopped. Since we have closed the channel there will be a channel close callback function which will be executed after the code goes into ioloop. In the channel close callback function ioloop is stopped.

One nice observation i had was that if i don't start ioloop here (at line 336 above), channel close callback will be executed in the future session when the same connection is used for making a separate channel and ioloop is started.